F3/feature/simdrivelabel

calbaker opened the pull request:
@michael-okeefe , I could sure use some help troubleshooting teh `test_net_accel_calc` failure in:
https://github.com/NREL/fastsim/actions/runs/16008814973/job/45161617012

As of 727ea7bf16f8df896e, it's
```
---- simdrivelabel::tests::test_net_accel_calc stdout ----

thread 'simdrivelabel::tests::test_net_accel_calc' panicked at fastsim-core/src/simdrivelabel.rs:1403:49:
called `Result::unwrap()` on an `Err` value: [fastsim-core/src/simdrivelabel.rs:23]

Caused by:
    0: [fastsim-core/src/simdrive.rs:338]
       time step: TrackedState(1, Fresh)
    1: [fastsim-core/src/simdrive.rs:439]
    2: [fastsim-core/src/vehicle/vehicle_model.rs:416]
    3: [fastsim-core/src/vehicle/conv.rs:94]
    4: [fastsim-core/src/vehicle/powertrain/fuel_converter.rs:285]
       `pwr_out_req` (23.307e3 W) must be < `self.state.pwr_prop_max` (21.050e3 W)
```

kylecarow commented:
@calbaker on [887f7aa](https://github.com/NREL/fastsim/pull/220/commits/887f7aa49fabb6843b6755abf038c2ca623e4156), we also did something similar in #194 BTW. We may be able to combine this code

Commit:
added more asserts

Commit:
Merge branch 'f3/simdrivelabel-partial' into f3/feature/simdrivelabel

Commit:
uncommented failing tests that need to be fixed

calbaker commented:
@michael-okeefe could you see if you can get this one over the finish line?  The three failing tests are checking against fastsim-2.

Commit:
Merge branch 'f3/simdrivelabel-partial' into f3/feature/simdrivelabel

Commit:
Merge branch 'f3/simdrivelabel-partial' into f3/feature/simdrivelabel

Commit:
Merge branch 'f3/simdrivelabel-partial' into f3/feature/simdrivelabel

Commit:
Merge branch 'f3/simdrivelabel-partial' into f3/feature/simdrivelabel

Commit:
Merge branch 'f3/simdrivelabel-partial' into f3/feature/simdrivelabel

Commit:
Merge branch 'f3/simdrivelabel-partial' into f3/feature/simdrivelabel

Commit:
Merge branch 'f3/simdrivelabel-partial' into f3/feature/simdrivelabel

Commit:
Merge branch 'f3/simdrivelabel-partial' into f3/feature/simdrivelabel

Commit:
Fix incorrect f2 variable reference for motor/controller efficiency settings.

Commit:
Ensure that the state variable i is less than the length of the cycle it is indexing

Commit:
Update powertrian type to be PlugInHybridElectricVehicle

Commit:
Switch to walk_once on PHEV to ensure SOC is being set; reset state variables

Commit:
Change to walk_once() to avoid iteration on accel

Commit:
Add (temporary) zeroing of aux loads.

With pwr_aux_base set to 0 kW, we can actually run the HEV label test without panic to the end. This shows that there must be a discrepancy in how aux loads are being handled at the "max power" phase vs the backward solution phase. This seems to only happen for the HEV.

Commit:
Enable hev to run without panic.

This checks the electric machine for whether
the maximum output is being requested (either
for tractive effort or regen). If so, the same
efficiency value used when calculating the max
possible effort is reused as the efficiency value.
This prevents slight lookup discrepancies and
allows the backward path to match the forward
path for the powertrain.

Commit:
Ensure efficiency is in valid range

Commit:
Fix incorrect J/kWh value.

Commit:
Call init again to ensure EM eff map is re-generated.

The eff_interp_at_max_input map was not being regenerated after the tweak done in f3veh_with_f2_eff(.) call. This corrects the issue.

Commit:
Pull out pieces of calculation for easier inspection.

Commit:
Correct detection of max output

To detect if we have max output or not, we check with pwr_mech_fwd_out_max (NOT pwr_mech_prop_out).

Commit:
Change lookup strategy for EM to linear.

This enables the BEV to pass.

Commit:
Merge branch 'fastsim-3' of github.com:NREL/fastsim into f3/feature/simdrivelabel

Commit:
intermediate debugging state

Commit:
Increase tolerance for conv.

This allows the conventional vehicle to pass the test without the f3veh_with_f2_eff() call.

Commit:
Remove f3veh_with_f2_eff()

After testing and discussing as a team, the value obtained from f3veh_with_f2_eff() was deemed not sufficient to keep it around (with its additional complexity). Therefore, we are removing it and opting to increase the f2 vs f3 tolerance (where needed) instead.

Commit:
Merge branch 'f3/feature/simdrivelabel' of https://github.com/NREL/fastsim into f3/feature/simdrivelabel

Commit:
Remove printout as it causes more tests to fail

Commit:
Correct error string.

Commit:
Update defaults for hybrid controls.

Commit:
Follow ruff format suggestion to add comma

Commit:
Update Prius model with updated defaults

Commit:
Hook up a checker for the RGWDB powertrain control.

This method does some sanity checking around the values used for the discharge, charge, and regen buffers and their associated speeds.

Commit:
Fix issue with mass query

Commit:
Add helper function for simdrive label fe

This helper function checks the fuel economy comparison between fastsim 2 and fastsim 3.

Commit:
Rework tolerance limits for accel.

Commit:
Add error and tolerance to output

For the simdrive label helper function, we now output the error vs the tolerance limit so it is easy to see what error value we have.

Commit:
Implement traction limits

Added code that limits the traction based on friction limitations for the tire/roadway interface.

michael-okeefe commented:
This post documents the current status of comparison between FASTSim 2 and FASTSim 3 over the vehicles used for comparison with the simdrivelabel functionality.

The vehicles used by powertrain are:

| Powertrain | Vehicle                                   |
| ---------- | ----------------------------------------- |
| CONV       | fastsim-2_2012_Ford_Fusion.yaml           |
| BEV        | fastsim-2_2022_Renault_Zoe_ZE50_R135.yaml |
| HEV        | fastsim-2_2016_TOYOTA_Prius_Two.yaml      |
| PHEV       | f2-vehicles/2016 CHEVROLET Volt.yaml      |

Note: values for EPA results are taken from the validation information from the yaml file when available (see `val_udds_mpgge` and `val_hwy_mpgge` and `val_comb_mpgge` and similar). Otherwise, we reference the following links:

* [2012 Ford Fusion](https://www.fueleconomy.gov/feg/bymodel/2012_ford_fusion.shtml)
* [2022 Renault Zoe ZE50](https://ev-database.org/car/1164/Renault-Zoe-ZE50-R110)
* [2016 Toyota Prius](https://www.fueleconomy.gov/feg/bymodel/2016_Toyota_Prius.shtml)
* [2016 Chevy Volt](https://www.fueleconomy.gov/feg/bymodel/2016_Chevrolet_Volt.shtml)

# City (UDDS)

| Powertrain   | FASTSim 2 | FASTSim 3 | EPA City |
| ------------ | --------- | --------- | -------- |
| CONV (mpg)   | 34.383    | 33.827    | 22       |
| BEV (kWh/mi) | 0.187     | 0.187     | N/A      | 
| HEV (mpg)    | 63.411    | 71.426    | 54 - 58  |
| PHEV (mpgge) | 354.666   | 370.064   | ??       |

# Highway (HWY)

| Powertrain   | FASTSim 2 | FASTSim 3 | EPA City |
| ------------ | --------- | --------- | -------- |
| CONV (mpg)   | 46.984    | 46.008    | 32       |
| BEV (kWh/mi) | 0.223     | 0.226     | N/A      | 
| HEV (mpg)    | 70.214    | 72.387    | 50 - 53  |
| PHEV (mpgge) | 273.412   | 282.759   | ??       |

# Combined

| Powertrain   | FASTSim 2 | FASTSim 3 | EPA City |
| ------------ | --------- | --------- | -------- |
| CONV (mpg)   | 39.102    | 38.402    | 25.6     |
| BEV (kWh/mi) | 0.203     | 0.204     | ??       | 
| HEV (mpg)    | 66.302    | 71.855    | 52 - 56  |
| PHEV (mpgge) | 312.830   | 324.919   | \*       |

\* Volt has 106 MPGe (electric) and 42 MPG (reg gas).

# Acceleration (0 to 60 mph)

Values are in seconds. The Google Search results are from a query of the form "What is the 0 to 60 mph acceleration time for the {{year}} {{make}} {{model}}?". Sport versions of the vehicle were eliminated from the results.

| Powertrain   | FASTSim 2 | FASTSim 3 | Google Search |
| ------------ | --------- | --------- | ------------- |
| CONV         | 8.246     | 8.598     | 7.3 - 8.9     |
| BEV          | 7.892     | 9.252     | 9.5           |
| HEV          | 7.817     | 7.997     | 9.8           |
| PHEV         | 7.165     | 7.963     | 8.4           |

Commit:
Update simdrive label comparison function

The function has been updated so that all results are reported if any of the tests fail to match within tolerance.

Commit:
Update tolerances to pass all tests.

Note: all tolerances for tests should be given a second look by the team. This commit just puts things in order should we want to move forward with the current PR as-is.

Commit:
Update saved results

Updated due to change in default speed_soc_fc_on_buffer and speed_soc_disch_buffer

Commit:
Add progress in separating simdrive label into a sim and post-processing function

Commit:
Fix issue with speeds and add tolerance.

The speed issue was that we were pulling in
the achieved speed in m/s instead of mph. For
the tolerance, we were comparing f64 values
for direct equality. Now we add a bit of tolerance
in to see that the absolute difference (or the
absolute difference fraction, (a - b)/b) is within
tolerance. We don't expect that individual
tolerances should need to be exposed for this
function because the fastsim 2 and 3 results
should be quite close. We just need to account
for some floating point error.

Commit:
Add check for BEV simdrive label calcs

Checks that the FASTSim 3 simdrive label
checks work for FASTSim 2 vehicle inputs and
that (roughly) the same outputs are calculated.

Commit:
Interim checkin with code passing tests.

Commit:
All tests passing on simdrive label comparisons

Commit:
Extract helper functions for running simdrive PHEV with an initial soc

Commit:
Finish separating get_label_fe into simulation and post-processing functions

Commit:
Remove commented code and rename function

Commit:
Rename function for clarity.

Commit:
Update function name for clarity

Commit:
Annotate functions with documentation

Commit:
Pull tolerance out to be by powertrain.

Since the PHEV requires much more tolerance (0.002 or 0.2%) versus the other powertrains, we pull that out to be explicit in the powertrain tests. This will hopefully signal future developers that there is some remaining (slight) disagreement between FASTSim 2 and FASTSim 3 with regard to label fuel economy calculation.

kylecarow commented on fastsim-core/src/simdrivelabel.rs:
```diff
@@ -17,10 +17,64 @@ fn first_grtr(arr: &[f64], cut: f64) -> Option<usize> {
+    if accel_data.speed_mph.iter().any(|&x| x >= 60.0) {
+        // Create interpolator from speed to time
+        let interp: InterpolatorEnumOwned<f64> = {
+            let wrapped_interp = InterpolatorEnum::new_1d(
```
@michael-okeefe little note here, I don't think there's a need for an `InterpolatorEnum` here (or any type annotations), a call to `Interp1D::new()` should be sufficient :)

The enum types are useful when we want the ability to change the underlying data (well, techincally, the dimensionality of the underlying data)

michael-okeefe replied:
Thank you, @kylecarow . I'll update it and push the change up.

michael-okeefe replied:
@kylecarow I just updated the PR with a change to simplify this. Thank you for the heads up!

Commit:
Simplify use of Interp1D.

Per comment from Kyle, simplify the usage of the Interp1D. Remove type that can be inferred on interp and simplify the constructor to just Interp1D and use the ndarray constructors explicitly.

robinsteuteville committed:
fixing the try_from() function translating fastsim-2 vehicles to fastsim-3 vehicles, so PHEVs are read in as PlugInHybridElectricVehicle and not HybridElectricVehicle

robinsteuteville committed:
my mistake -- thought that the PHEVs were being read in wrong but they actually should be read into the HybridElectricVehicle struct -- since PHEVs and HEVs share the same underlying fastism-3 structure
